majority-multisign/src/MajorityMultiSign/OnChain.hs

```diff
@@ -69,19 +69,28 @@ hasNewSignatures (UpdateKeysAct keys) MajorityMultiSignDatum {signers} ctx =
 {-# INLINEABLE hasCorrectToken #-}
 hasCorrectToken :: MajorityMultiSignValidatorParams -> ScriptContext -> MajorityMultiSignDatum -> Bool
 hasCorrectToken MajorityMultiSignValidatorParams {asset} ctx expectedDatum =
-  traceIfFalse "Couldn't find asset" (isJust assetTxOut)
-    && traceIfFalse
-      "Incorrect output datum"
-      ((assetTxOut >>= txOutDatumHash) == findDatumHash (Datum $ PlutusTx.toBuiltinData expectedDatum) (scriptContextTxInfo ctx))
+  case result of
+    Left errMsg -> trace errMsg False
+    Right () -> True
   where
     continuing :: [TxOut]
     continuing = Ledger.getContinuingOutputs ctx
 
     checkAsset :: TxOut -> Maybe TxOut
     checkAsset txOut = if assetClassValueOf (txOutValue txOut) asset > 0 then Just txOut else Nothing
 
-    assetTxOut :: Maybe TxOut
-    assetTxOut = firstJust checkAsset continuing
+    justOrErr :: BuiltinString -> Maybe a -> Either BuiltinString a
+    justOrErr errMsg = maybe (Left errMsg) Right
+
+    result :: Either BuiltinString ()
+    result = do
+      assetTxOut <- justOrErr "Couldn't find asset" $ firstJust checkAsset continuing
+      datumHash <- justOrErr "Continuing output does not have datum" $ txOutDatumHash assetTxOut
+      expectedDatumHash <-
```

Comment:
could potentially use `liftA2` here to save so much binding

Reply:
Do you mean like this? https://github.com/mlabs-haskell/majority-multisign/pull/40/commits/b394e88d0c139eb08ac5d6afe2487f9c008f092c#diff-05b0faef073b02efdaf6ef1f3e840f860ad75e580fe6ba4dcef9ff7a45e12327R91